File: server/src/main/java/ru/fullrest/mfr/server/service/TelegramUserService.java
package ru.fullrest.mfr.server.service;

import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import ru.fullrest.mfr.server.configuration.CacheConfiguration;
import ru.fullrest.mfr.server.model.entity.TelegramUser;
import ru.fullrest.mfr.server.model.repository.TelegramUserRepository;

import javax.transaction.Transactional;

@Service
@RequiredArgsConstructor
public class TelegramUserService {

    private final TelegramUserRepository telegramUserRepository;

    @Cacheable(value = CacheConfiguration.USER_CACHE,
            key = "T(ru.fullrest.mfr.server.configuration.CacheConfiguration).USER_CACHE + '|' + #userId",
            unless = "#result == null")
    public TelegramUser findById(Integer userId) {
        return telegramUserRepository.findById(userId).orElse(null);
    }

    @Cacheable(value = CacheConfiguration.USER_CACHE,
            key = "#root.methodName",
            unless = "#result == null")
    public Iterable<TelegramUser> findAll() {
        return telegramUserRepository.findAll();
    }

    @Transactional
    @CacheEvict(value = CacheConfiguration.USER_CACHE, allEntries = true)
    public void save(TelegramUser user) {
        telegramUserRepository.save(user);
    }

    @Transactional
    @CacheEvict(value = CacheConfiguration.USER_CACHE, allEntries = true)
    public void delete(TelegramUser user) {
        telegramUserRepository.delete(user);
    }
}

File: server/src/main/kotlin/com/lezenford/mfr/server/service/model/TelegramUserService.kt
package com.lezenford.mfr.server.service.model

import com.lezenford.mfr.server.configuration.CacheConfiguration
import com.lezenford.mfr.server.model.entity.TelegramUser
import com.lezenford.mfr.server.model.repository.TelegramUserRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class TelegramUserService(
    private val telegramUserRepository: TelegramUserRepository
) {
    @Cacheable(value = [CacheConfiguration.USER_CACHE], unless = "#result == null")
    fun findById(userId: Long): TelegramUser? {
        return telegramUserRepository.findById(userId).orElseGet { null }
    }

    @Cacheable(value = [CacheConfiguration.USER_CACHE], unless = "#result.isEmpty()")
    fun findAll(): List<TelegramUser> {
        return telegramUserRepository.findAll().toList()
    }

    @Transactional
    @CacheEvict(value = [CacheConfiguration.USER_CACHE], allEntries = true)
    fun save(user: TelegramUser) {
        telegramUserRepository.save(user)
    }

    @Transactional
    @CacheEvict(value = [CacheConfiguration.USER_CACHE], allEntries = true)
    fun deleteById(id: Long) {
        telegramUserRepository.deleteById(id)
    }
}